### ISSUE TITLE
running windows code in net6 target project

### ISSUE THREAD

**Issue by MiroKov** [dotnet/runtime#86797]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

![image](https://github.com/dotnet/aspnetcore/assets/18582187/7c4c3730-1782-4f92-a9cf-5c2a27536512)
![image](https://github.com/dotnet/aspnetcore/assets/18582187/fd5ebee4-8361-43e8-aeb7-79318a4fb7dd)

This code produces the following error:
![image](https://github.com/dotnet/aspnetcore/assets/18582187/b81a17db-d47a-4e6a-adfe-3845e3b638f5)
This used to work before (for instance in .net core 3.1), but not anymore it seems. To me it makes no sense, since I am running the code in windows, and I am testing accordingly.

### Expected Behavior

I expect the code to run.

### .NET Version

7.0.100

**Comment by vitek-karas** [comment]
Is the application published as self-contained or not?
Can you look in the debugger where did it load the `System.Drawing.Common` from?
What is the version of the System.Drawing.Common in the project?
Is your application targeting .NET 7 or some other version?

**Comment by huoyaoyuan** [comment]
Note: you can now use `System.OperatingSystem.IsWindows` to check OS, and remove the `#if`.

**Comment by MiroKov** [comment]
Hi, this is the system.drawing.common that is referenced by the project:
![image](https://github.com/dotnet/runtime/assets/18582187/85c1bce7-725a-4bd4-a403-bcc555ccc628)
The application is published as framework-dependant, but the error happens while in debug:
![image](https://github.com/dotnet/runtime/assets/18582187/805cd268-18fd-450b-a19b-9d28d571cb82)
Application is targeting net6.

**Comment by vitek-karas** [comment]
For some reason the binder resolved the assembly to the "fallback" version - which throws PNS for everything.
I assume there's a package reference to `System.Drawing.Common` in your project somewhere (or transitively through something).

Could you please open the `app.deps.json` file for your app and find all sections relevant to `System.Drawing.Common` (just search for that string). There should be a section which look something like:
```JSON
"System.Drawing.Common/6.0.0": {
  "runtime": {
  }
}
```

Could you please share what it looks like?

You can also turn on [host tracing](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-environment-variables#corehost_trace) - basically `COREHOST_TRACE=1` and `COREHOST_TRACE_FILE=host.txt`. Note that the file contains lot of information about your machine and user (local environment). But again it should be easy to find the parts about `System.Drawing.Common.dll`. It would be interesting to see those as well.

**Comment by MiroKov** [comment]
Here are all Szstem.Drawing.Common that exist in the file:
![image](https://github.com/dotnet/runtime/assets/18582187/ac5079dd-f2e6-465e-b477-de2a653a59f7)
![image](https://github.com/dotnet/runtime/assets/18582187/b3d60fab-6df2-42a6-9d49-644f58a02bc2)
![image](https://github.com/dotnet/runtime/assets/18582187/3fefe75a-4dd6-4dd9-831b-ed81393b30be)
![image](https://github.com/dotnet/runtime/assets/18582187/eba94f0a-1591-414c-867d-98f6f87bd5e1)
Sorry, don't know why, but I do not get the host.txt created, don't know what I'm doing wrong:
![image](https://github.com/dotnet/runtime/assets/18582187/cf6ed788-805a-47b4-95be-4193612914a7)

**Comment by vitek-karas** [comment]
If you're running in PowerShell you have to setup environment variables like:
```
$env:COREHOST_TRACE=1
$env:COREHOST_TRACE_FILE="host.txt"
```

Thanks for the `.deps.json` but that looks the way I would expect. Maybe the trace will tell us more.

**Comment by MiroKov** [comment]
hi,
the host.txt is simply not created. I even set the variables programmatically on the beginning of Program.cs with:
![image](https://github.com/dotnet/runtime/assets/18582187/5bacd823-671f-4f5d-a699-1e064ba05320)

and when debugging, it says:
![image](https://github.com/dotnet/runtime/assets/18582187/2a02b03f-257a-4ebd-93be-d092fb6fc7ad)

However, no host.txt exists.

**Comment by vitek-karas** [comment]
That is really weird.
Setting the env. variables in Program.cs will not do anything, the host runs before your program gets to main. In powershell something like this should work:
```
$env:COREHOST_TRACE=1
$env:COREHOST_TRACE_FILE="host.txt"
& myapp.exe
```

The `host.txt` will be created in the current directory.
Do you run the app directly as an executable, or you're using some other way to launch it?

**Comment by MiroKov** [comment]
I tried it now with dotnet run in the debug folder, this came out:
![image](https://github.com/dotnet/runtime/assets/18582187/becdbc4d-7251-4e01-ab31-b603da42f2dc)
...
![image](https://github.com/dotnet/runtime/assets/18582187/91713f83-96a4-4c5e-8dea-a08540119277)
..
![image](https://github.com/dotnet/runtime/assets/18582187/f74d88d9-7530-4a1c-be0f-9c18de11bfac)
..
![image](https://github.com/dotnet/runtime/assets/18582187/99d49d71-b6d8-4581-8c6f-28dd21c76163)
..
![image](https://github.com/dotnet/runtime/assets/18582187/bc73c47d-43cf-4ac5-8d26-565cacec5339)

**Comment by vitek-karas** [comment]
Thanks - that one looks correct.
But that's not the same as the failure you're seeing.
The resolved path in this case is:
`C:\Program Files\dotnet\....\win\lib\net6.0\System.Drawing.Common.dll`

But in your failure (from the debugger) the path is:
`C:\Users\...\.nuget\...\lib\net6.0\System.Drawing.Common.dll`

The important difference is the lack of `win` subdirectory in the failure case.

Does your app use plugins or something similar, basically anything which would try to load assemblies dynamically at runtime?
Out of curiosity - in the run where the host trace worked, did it fail the same way as in the issue description above?

**Comment by MiroKov** [comment]
Yes, we use plugins. Basically, foreach plugin we load the references that we find in the plugin directory that were not loaded in the default context with the same assembly full name.

Regarding your other question, I have an angular app that I managed to run only in Rider with IIS express configuration, something with the proxy doesn't seem to work when using dotnet run - cannot see the web site in the browser.

What I find strange, is this folder underlined with yellow:
![image](https://github.com/dotnet/runtime/assets/18582187/d996ca50-2f95-4ba6-8106-987e824db2ac)
Where does this come from, and is this maybe the reason for error?

**Comment by elinor-fung** [comment]
re: tracing - The environment variable for writing the trace to a file is `COREHOST_TRACEFILE` (only one underscore).

> Basically, foreach plugin we load the references that we find in the plugin directory that were not loaded in the default context with the same assembly full name.

What extension point(s) are you using to do this - for example, `AssemblyDependencyResolver`, `AssemblyLoadContext.Resolving` event? And which `AssemblyLoadContext` are you loading them into - default ALC versus separate `AssemblyLoadContext` instances or subclasses?

> Where does this come from, and is this maybe the reason for error?

The path logged as the 'component main assembly path' would be the path you pass to the [constructor of `AssemblyDependencyResolver`](https://learn.microsoft.com/dotnet/api/system.runtime.loader.assemblydependencyresolver.-ctor).

**Comment by vitek-karas** [comment]
Sorry about the wrong environment variable name... my bad :-(

If you're using `AssemblyDependencyResolver` (which seems to be the case based on the last trace) then it would be interesting to reproduce the failure with the tracing on and look for the resolved assembly path in the trace - it's probably there somewhere (unless you have additional resolution customization code). The trace will include even the resolution happening on behalf of `AssemblyDependencyResolver`.

**Comment by MiroKov** [comment]
Here are all the mentions:
![image](https://github.com/dotnet/runtime/assets/18582187/d3be7f84-640b-4cb3-9149-3cdb985f0ce0)
![image](https://github.com/dotnet/runtime/assets/18582187/fc70a115-71a8-437f-b349-b1a581d81e60)
![image](https://github.com/dotnet/runtime/assets/18582187/65995897-da75-4840-82d0-51dd7e9bbe05)
![image](https://github.com/dotnet/runtime/assets/18582187/cf677bc0-dabf-4c2e-bd97-9da423287531)
![image](https://github.com/dotnet/runtime/assets/18582187/d207f76f-afc7-4c84-a12c-bf679383bc9c)
![image](https://github.com/dotnet/runtime/assets/18582187/a7cef898-5ff4-4e65-973a-5888cf998c5a)
![image](https://github.com/dotnet/runtime/assets/18582187/e1dcd732-f36f-4bbe-9bb5-ac1cad5d3175)
![image](https://github.com/dotnet/runtime/assets/18582187/0521db11-337d-4c6b-a9e3-cf541cb0d9ef)
![image](https://github.com/dotnet/runtime/assets/18582187/8173f580-5a23-432d-8ec0-47adaa625545)
![image](https://github.com/dotnet/runtime/assets/18582187/7948c8ca-52d8-4987-b857-69559ecc30fb)
![image](https://github.com/dotnet/runtime/assets/18582187/e153e9ed-b368-482d-90f4-12749bc2ee9e)
![image](https://github.com/dotnet/runtime/assets/18582187/10f0f054-7558-48a4-9ab6-9af0cc9a97c1)
![image](https://github.com/dotnet/runtime/assets/18582187/a130bf7d-1f48-4b75-9c17-00fab1a7469f)
![image](https://github.com/dotnet/runtime/assets/18582187/e9ff36bf-665f-44b1-9bcf-b62ff871386f)
![image](https://github.com/dotnet/runtime/assets/18582187/f4fb028c-74b7-4114-9a96-713f49be997d)
![image](https://github.com/dotnet/runtime/assets/18582187/a18482a2-23fc-4f1a-974e-08088ce4cc85)
![image](https://github.com/dotnet/runtime/assets/18582187/1a39e4a1-822b-46f0-a7cc-481a4a7d23fe)
![image](https://github.com/dotnet/runtime/assets/18582187/8119fc96-afa8-4197-b7b8-e020cb048e68)
![image](https://github.com/dotnet/runtime/assets/18582187/d9760a46-6bde-4c37-a44d-1259063612c9)
![image](https://github.com/dotnet/runtime/assets/18582187/681d54d2-3127-4876-9217-7ae2440d40fd)
![image](https://github.com/dotnet/runtime/assets/18582187/7470fa42-f206-4510-a871-33a483367ff1)
![image](https://github.com/dotnet/runtime/assets/18582187/a52e1154-a570-4c8e-a6c4-c59beb8f8b0e)
![image](https://github.com/dotnet/runtime/assets/18582187/2b391e72-eae4-4b8d-96dc-ae0a664c1098)
![image](https://github.com/dotnet/runtime/assets/18582187/d4b29e01-4612-4f27-935d-1a21f02727e7)
![image](https://github.com/dotnet/runtime/assets/18582187/5958692d-f829-45ca-8db2-10782c5df58c)
![image](https://github.com/dotnet/runtime/assets/18582187/cd92100b-fd5a-4d4b-a4a4-f4c7acae4c50)
![image](https://github.com/dotnet/runtime/assets/18582187/bba1ef1d-0481-40ad-8b9f-807cad9b9852)
![image](https://github.com/dotnet/runtime/assets/18582187/8367593e-619c-4a86-b173-92fdde5f013c)
![image](https://github.com/dotnet/runtime/assets/18582187/b99b8f77-c48e-4d13-b98b-effaa4d9c963)
![image](https://github.com/dotnet/runtime/assets/18582187/f08b6173-929c-48f5-882e-cc163ce342fa)
![image](https://github.com/dotnet/runtime/assets/18582187/50ce853a-52e8-4fb6-8d0e-5b06972ca6b4)

 
> What extension point(s) are you using to do this - for example, `AssemblyDependencyResolver`, `AssemblyLoadContext.Resolving` event? And which `AssemblyLoadContext` are you loading them into - default ALC versus separate `AssemblyLoadContext` instances or subclasses?
> 
We use `assemblyLoadContext.LoadFromAssemblyPath(referenceFullPath)` to load the assemblies in each plugin context.
The extension point we use is the AssemblyDependencyResolver.
> 
> The path logged as the 'component main assembly path' would be the path you pass to the [constructor of `AssemblyDependencyResolver`](https://learn.microsoft.com/dotnet/api/system.runtime.loader.assemblydependencyresolver.-ctor).

Here is the code for the PluginLoadContext:
`

    internal class PluginLoadContext : AssemblyLoadContext
    {
        private readonly AssemblyDependencyResolver resolver;

        public PluginLoadContext(string pluginDirectory) : base(pluginDirectory)
        {
            resolver = new AssemblyDependencyResolver(pluginDirectory);
        }

        protected override Assembly Load(AssemblyName assemblyName)
        {
            var assemblyPath = resolver.ResolveAssemblyToPath(assemblyName);
            var pluginAssembly = assemblyPath != null ? LoadFromAssemblyPath(assemblyPath) : null;
            return pluginAssembly;
        }

        protected override IntPtr LoadUnmanagedDll(string unmanagedDllName)
        {
            var libraryPath = resolver.ResolveUnmanagedDllToPath(unmanagedDllName);
            return libraryPath != null ? LoadUnmanagedDllFromPath(libraryPath) : IntPtr.Zero;
        }
    }
`
Here, the path was something like:
`c:\Users\...\bin\Debug\.\Plugins\Common` after which I changed this to:
`c:\Users\...\bin\Debug\Plugins\Common`

but the error from the beginning stayed the same.

**Comment by vitek-karas** [comment]
I don't see anything problematic in the traces.
I would try to modify the plugin load context and maybe in a debugger stop on every load of the affected dll and check what path it's loading it from.

The runtime itself should not come up with a NuGet cache path on its own... it's weird.
Could you check if there is no `<name>.runtimeconfig.dev.json` file involved? It would typically contain something like this in it:
```JSON
{
  "runtimeOptions": {
    "additionalProbingPaths": [
      "C:\\Users\\#####\\.dotnet\\store\\|arch|\\|tfm|",
      "C:\\Users\\#####\\.nuget\\packages",
      "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\NuGetPackages"
    ]
  }
}
```

Which would be the way for runtime to learn about the nuget cache location. This could also possibly hint which component is causing the problem.

**Comment by MiroKov** [comment]
> I don't see anything problematic in the traces. I would try to modify the plugin load context and maybe in a debugger stop on every load of the affected dll and check what path it's loading it from.

Here is the assemblyPath from AssemblyDependencyResolver (it is the same path that goes to the PluginContext):
![image](https://github.com/dotnet/runtime/assets/18582187/23dd574a-2600-46e3-9bee-9faf7cd80e54)
and the file that gets loaded is:
![image](https://github.com/dotnet/runtime/assets/18582187/395144ed-3aac-4b97-a868-a91de2f04c14)
which is:
![image](https://github.com/dotnet/runtime/assets/18582187/c5b6b193-3c5c-4c21-8d28-d89b2956b232)

Do I understand this correctly, the problem is because this file is bad (contains errors)?. 

> The runtime itself should not come up with a NuGet cache path on its own... it's weird. Could you check if there is no `<name>.runtimeconfig.dev.json` file involved? It would typically contain something like this in it:
> 
> ```json
> {
>   "runtimeOptions": {
>     "additionalProbingPaths": [
>       "C:\\Users\\#####\\.dotnet\\store\\|arch|\\|tfm|",
>       "C:\\Users\\#####\\.nuget\\packages",
>       "C:\\Program Files (x86)\\Microsoft Visual Studio\\Shared\\NuGetPackages"
>     ]
>   }
> }
> ```
> 
There are no such files. Only .runtimeconfig.json, but they look like this:
![image](https://github.com/dotnet/runtime/assets/18582187/a5c038d8-358d-403e-9c67-b5fab3456397)

They have no "additionalProbingPaths".

**Comment by vitek-karas** [comment]
> Do I understand this correctly, the problem is because this file is bad (contains errors)?.

It's not "bad" as such. The System.Drawing.Common only works on certain platforms, but the API is available on all platforms. So the package contains 3 files:
* Windows version - which works and that's the one you want
* Linux version - which also works, but obviously can't be used on Windows
* Fallback version - which is used on all other platforms and it throws `PlatformNotSupportedException` from all APIs

What you're seeing is that for some reason the fallback version is used even though it's running on Windows.

You can't tell which one is which by looking at file properties unfortunately. You would have to use tools like ILSpy to look inside (if the method you're calling throws or not), or guess by size, the fallback version should be smaller since it actually doesn't contain any real code in it.

Given that the file comes from the `Plugin/Common` directory I would look into how is that constructed:
* What creates the files in that directory? Assuming it's some project, how do you build it, what TFM it targets....
* You pass a path to the `AssemblyDependencyResolver` which points to a `.dll`, look next to it for `.deps.json` file and then look inside for any mentions of `System.Drawing.Common` - maybe this is were the problem is.

**Comment by vitek-karas** [comment]
> the way I understand it, in this case will PluginContext use the System.Drawing.Common.dll from default context

Yes - that is correct.

Would be really interesting to see the trace for the call to the `AssemblyDependencyResolver` which returned the path to the wrong assembly. It would be in the:
```
corehost_resolve_component_dependencies results: {
  assembly_paths: 'path to the wrong assembly'
}
```

That will be at the end of the interesting section. The beginning is the preceding occurrence of `corehost_resolve_component_dependencies`. (It should say something like "Invoke ... ").

**Comment by MiroKov** [comment]
Hi,
I haven't found the section with the path. [Here](https://github.com/MiroKov/SpaceHolder/blob/main/host-sanitized.txt) is the trace file, maybe you can find it?

**Comment by vitek-karas** [comment]
That is a trace of `dotnet run`, it doesn't contain the app itself as far as I can tell. It only contains `dotnet run` and then two invocations of msbuild. It's possible that there will be another `host.txt` file in the directory with the app.

Alternatively I would suggest to run the app directly - either via `app.exe` or `dotnet app.dll`. That should give you a trace of only the app and nothing else.

**Comment by elinor-fung** [comment]
> Here, the pluginDirectory path was something like:
> `c:\Users\...\bin\Debug\.\Plugins\Common`
>
> You pass a path to the `AssemblyDependencyResolver` which points to a `.dll`, look next to it for `.deps.json` file 

The path should point to the actual assembly (the .dll path as Vitek mentioned), not the containing directory.

We look for a .deps.json based on that path. If you just pass in the directory, we will look for a file next to that directory with the same name. From the log:
> Could not locate the dependencies manifest file [C:\bin\Debug\Plugins\Common.deps.json].

If there is no .deps json file, everything in that directory - in this case `C:\bin\Debug\Plugins\` - is treated as an assembly dependency and the directory is added to the native / resource search paths, resulting in:
```
corehost_resolve_component_dependencies results: {
  assembly_paths: 'C:\bin\Debug\Plugins\Common;'
  native_search_paths: 'C:\bin\Debug\Plugins\;'
  resource_search_paths: 'C:\bin\Debug\Plugins\;'
}
```

Can you switch your usage of `AssemblyDependencyResolver` to point to the `.dll` instead of the directory?

All that said, I still don't know why `System.Drawing.Common` would be resolved by that `AssemblyDependencyResolver` - I would expect it to simply not find it, since it is not in the assembly paths.

> and the file that gets loaded is:
> `string assemblyPath = Path.Combine(parent.FullName, path2);`

Where are `parent.FullName` and `path2` from here?

Another thing that could be helpful is collecting a [trace with assembly loading info](https://learn.microsoft.com/dotnet/core/dependency-loading/collect-details).

Thanks for bearing with us on all the info collection here.

**Comment by MiroKov** [comment]
> > Here, the pluginDirectory path was something like:
> > `c:\Users\...\bin\Debug\.\Plugins\Common`
> > You pass a path to the `AssemblyDependencyResolver` which points to a `.dll`, look next to it for `.deps.json` file
> 
> The path should point to the actual assembly (the .dll path as Vitek mentioned), not the containing directory.
> 
> We look for a .deps.json based on that path. If you just pass in the directory, we will look for a file next to that directory with the same name. From the log:
> 
> > Could not locate the dependencies manifest file [C:\bin\Debug\Plugins\Common.deps.json].
> 
> If there is no .deps json file, everything in that directory - in this case `C:\bin\Debug\Plugins\` - is treated as an assembly dependency and the directory is added to the native / resource search paths, resulting in:
> 
> ```
> corehost_resolve_component_dependencies results: {
>   assembly_paths: 'C:\bin\Debug\Plugins\Common;'
>   native_search_paths: 'C:\bin\Debug\Plugins\;'
>   resource_search_paths: 'C:\bin\Debug\Plugins\;'
> }
> ```

The way I see it, the path is wrong, what gets passed into it is:
`
C:\Users\Userx\source\repos\AppName\AppName\AppName\bin\Debug\Plugins\Common
`
![image](https://github.com/dotnet/runtime/assets/18582187/82a16624-2d40-495b-8521-ca49587c6ec0)


> Can you switch your usage of `AssemblyDependencyResolver` to point to the `.dll` instead of the directory?
> 
> All that said, I still don't know why `System.Drawing.Common` would be resolved by that `AssemblyDependencyResolver` - I would expect it to simply not find it, since it is not in the assembly paths.
> 
> > and the file that gets loaded is:
> > `string assemblyPath = Path.Combine(parent.FullName, path2);`
> 
> Where are `parent.FullName` and `path2` from here?

![image](https://github.com/dotnet/runtime/assets/18582187/65e4fd7f-8e21-4ca5-ae97-f6e454b46240)
So, the reference.Name is the assembly that was referenced (in this case System.Drawing.Common),
and the parent is the directory of the plugin which in this case is:

`
C:\Users\Userx\source\repos\AppName\AppName\AppName\bin\Debug\Plugins\Common
`

Basically, we look for the refereces of the plugin - in this case the plugin is Common.Data.dll, and then we load all references present in the directory into the corresponding PluginContext, but only if the reference with the same full name doesn't already exist in the Default load context.

> Another thing that could be helpful is collecting a [trace with assembly loading info](https://learn.microsoft.com/dotnet/core/dependency-loading/collect-details).
> 
> Thanks for bearing with us on all the info collection here.

**Comment by vitek-karas** [comment]
> he way I see it, the path is wrong, what gets passed into it is:
`C:\Users\Userx\source\repos\AppName\AppName\AppName\bin\Debug\Plugins\Common`

As Elinor pointed out the argument for the `AssemblyDependencyResolver` .ctor is expected to point to a file, specifically the `assembly.dll` of the plugin you're trying to load. Passing in a directory will lead to weird behavior.
https://learn.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblydependencyresolver.-ctor?view=net-7.0

> Basically, we look for the refereces of the plugin - in this case the plugin is Common.Data.dll, and then we load all references present in the directory into the corresponding PluginContext, but only if the reference with the same full name doesn't already exist in the Default load context.

That would work if all the plugins were built/published for the target RID only. Some libraries (like `System.Drawing.Common`) carry different assets for different target platforms (RIDs). If you build a library project without specifying the target RID the build will include all available assets and put them into `runtime` subdirectory. It then writes information into the `plugin.deps.json` file which tells the runtime which assets to use for which RID.

This will only work if your code calls `AssemblyDependencyResolver` with the plugin dll path, so that it can find the `plugin.deps.json` next to it. And it will also only work if the code doesn't preload all files in a given directory - not all those files may be the right asset for the current RID (as is the case here, the `System.Drawing.Common.dll` in the root directory is NOT the right asset for Windows).

> but only if the reference with the same full name doesn't already exist in the Default load context.

That is not the default behavior shown in the plugin samples - and the code you shared above for you `AssemblyLoadContext` doesn't do that either. The samples will isolate plugin dependencies, so if the plugin carries an assembly with it, that will be preferred over the from the host. If you want different behavior you need to modify the `AssemblyLoadContext.Load` method to prefer assemblies from the `Default` context (so something like calling `AssemblyLoadContext.Default.LoadFromAssemblyName` first).

**Comment by MiroKov** [comment]
> That is not the default behavior shown in the plugin samples - and the code you shared above for you `AssemblyLoadContext` doesn't do that either. The samples will isolate plugin dependencies, so if the plugin carries an assembly with it, that will be preferred over the from the host. If you want different behavior you need to modify the `AssemblyLoadContext.Load` method to prefer assemblies from the `Default` context (so something like calling `AssemblyLoadContext.Default.LoadFromAssemblyName` first).

I understand now, so for those packages / projects that we want to use from the DefaultLoadContext we need to set:
`
 <Private>false</Private>
`
`
 <ExcludeAssets>runtime</ExcludeAssets>
`
 in package/project reference element in .csproj.
